File: core/src/main/java/org/predicode/predicator/grammar/CharClass.java
package org.predicode.predicator.grammar;

import javax.annotation.Nonnull;

import static org.predicode.predicator.grammar.CodePoints.BACKSLASH;
import static org.predicode.predicator.grammar.NameCharPrinter.*;
import static org.predicode.predicator.grammar.NameCharPrinter.NAME_ESCAPE_PRINTER;


public enum CharClass {

    START_CHAR_CLASS(NAME_START_PRINTER, true, true) {
        @Override
        boolean nameGluesWith(@Nonnull CharClass next) {
            return true;
        }
    },
    LETTER_CHAR_CLASS(NAME_BODY_PRINTER, true, true) {
        @Override
        boolean nameGluesWith(@Nonnull CharClass next) {
            return super.nameGluesWith(next) && next != NUMERIC_CHAR_CLASS;
        }
    },
    NUMERIC_CHAR_CLASS(NAME_BODY_PRINTER, false, true) {
        @Override
        boolean nameGluesWith(@Nonnull CharClass next) {
            return super.nameGluesWith(next) && next != LETTER_CHAR_CLASS;
        }
    },
    OPERATOR_CHAR_CLASS(NAME_BODY_PRINTER, false, false),
    CONNECTOR_CHAR_CLASS(NAME_BODY_PRINTER, false, false),
    SEPARATOR_CHAR_CLASS(NAME_SEPARATOR_PRINTER, true, true) {
        @Override
        boolean nameGluesWith(@Nonnull CharClass next) {
            return true;
        }
    },
    BUILTIN_CHAR_CLASS(NAME_ESCAPE_PRINTER, false, false),
    QUOTE_CHAR_CLASS(NAME_ESCAPE_PRINTER, false, false),
    ESCAPE_CHAR_CLASS(NAME_ESCAPE_PRINTER, false, false),
    PUNCTUATION_CHAR_CLASS(NAME_ESCAPE_PRINTER, false, false),
    OTHER_CHAR_CLASS(NAME_CODE_PRINTER, false, false);

    @Nonnull
    public static CharClass of(int codePoint) {
        if (Character.isWhitespace(codePoint)) {
            return SEPARATOR_CHAR_CLASS;
        }
        if (OPERATORS.get(codePoint)) {
            return OPERATOR_CHAR_CLASS;
        }
        if (QUOTES.get(codePoint)) {
            return QUOTE_CHAR_CLASS;
        }
        if (BUILTINS.get(codePoint)) {
            return BUILTIN_CHAR_CLASS;
        }
        if (codePoint == BACKSLASH) {
            return ESCAPE_CHAR_CLASS;
        }

        final int type = Character.getType(codePoint);

        if (type >= Character.UPPERCASE_LETTER && type <= Character.OTHER_LETTER) {
            return LETTER_CHAR_CLASS;
        }
        if (type >= Character.DECIMAL_DIGIT_NUMBER && type <= Character.LETTER_NUMBER) {
            return NUMERIC_CHAR_CLASS;
        }
        if (type >= Character.MATH_SYMBOL && type <= Character.OTHER_SYMBOL) {
            return OPERATOR_CHAR_CLASS;
        }
        if (type == Character.DASH_PUNCTUATION || type == Character.CONNECTOR_PUNCTUATION) {
            return CONNECTOR_CHAR_CLASS;
        }
        if (type >= Character.START_PUNCTUATION && type <= Character.OTHER_PUNCTUATION) {
            return PUNCTUATION_CHAR_CLASS;
        }
        if (type >= Character.SPACE_SEPARATOR && type <= Character.PARAGRAPH_SEPARATOR) {
            return SEPARATOR_CHAR_CLASS;
        }

        return OTHER_CHAR_CLASS;
    }

    @Nonnull
    final NameCharPrinter printer;

    private final boolean nameStart;
    private final boolean nameEnd;

    CharClass(
            @Nonnull NameCharPrinter printer,
            boolean nameStart,
            boolean nameEnd) {
        this.printer = printer;
        this.nameStart = nameStart;
        this.nameEnd = nameEnd;
    }

    public final boolean nameStart() {
        return this.nameStart;
    }

    public final boolean nameEnd() {
        return this.nameEnd;
    }

    public final boolean separating() {
        return this == SEPARATOR_CHAR_CLASS;
    }

    boolean nameGluesWith(@Nonnull CharClass next) {
        return next != this;
    }

    @Nonnull
    CharClass printName(@Nonnull NamePrinter out, int codePoint) {

        final CharClass next = of(codePoint);

        final CharClass appender = this.printer.append(this, out, next);

        appender.printer.print(appender, out, next, codePoint);

        return appender;
    }

}

File: core/src/main/kotlin/org/predicode/predicator/grammar/CharClass.kt
package org.predicode.predicator.grammar

import java.util.*


private val OPERATORS: BitSet = BitSet(128).apply {
    set('%'.toInt())
    set('&'.toInt())
    set('|'.toInt())
    set('+'.toInt())
    set('-'.toInt())
    set('*'.toInt())
    set('/'.toInt())
    set('>'.toInt())
    set('<'.toInt())
    set('='.toInt())
    set('~'.toInt())
    set('^'.toInt())
}

private val BUILTINS: BitSet = BitSet(128).apply {
    set(':'.toInt())
    set('#'.toInt())
    set('$'.toInt())
    set('@'.toInt())
}

private val QUOTES: BitSet = BitSet(128).apply {
    set(BACKTICK)
    set(SINGLE_QUOTE)
    set(DOUBLE_QUOTE)
    set(UNDERSCORE)
}

enum class CharClass(
        private val printer: NameCharPrinter,
        val nameStart: Boolean = false,
        val nameEnd: Boolean = false) {

    START_CHAR_CLASS(NameStartPrinter, nameStart = true, nameEnd = true) {
        override fun nameGluesWith(next: CharClass) = true
    },
    LETTER_CHAR_CLASS(NameBodyPrinter, nameStart = true, nameEnd = true),
    NUMERIC_CHAR_CLASS(NameBodyPrinter, nameEnd = true),
    OPERATOR_CHAR_CLASS(NameBodyPrinter),
    CONNECTOR_CHAR_CLASS(NameBodyPrinter),
    SEPARATOR_CHAR_CLASS(NameSeparatorPrinter, nameStart = true, nameEnd = true) {
        override fun nameGluesWith(next: CharClass) = true
    },
    BUILTIN_CHAR_CLASS(NameEscapePrinter),
    QUOTE_CHAR_CLASS(NameEscapePrinter),
    ESCAPE_CHAR_CLASS(NameEscapePrinter),
    PUNCTUATION_CHAR_CLASS(NameEscapePrinter),
    OTHER_CHAR_CLASS(NameCodePrinter);

    val separating: Boolean
        get() = this == SEPARATOR_CHAR_CLASS

    open fun nameGluesWith(next: CharClass) = next != this

    internal fun printName(out: NamePrinter, codePoint: CodePoint): CharClass =
            of(codePoint).let { next ->
                printer.append(this, out, next).also { appender ->
                    appender.printer.print(appender, out, next, codePoint)
                }
            }

    private interface NameCharPrinter {

        fun append(current: CharClass, out: NamePrinter, next: CharClass): CharClass

        fun print(current: CharClass, out: NamePrinter, next: CharClass, codePoint: CodePoint)

    }

    private object NameStartPrinter : NameCharPrinter {

        override fun append(current: CharClass, out: NamePrinter, next: CharClass) =
                if (next.separating) current // Skip leading separators
                else {
                    // Open quote if the name does not start with allowed start symbol
                    if (out.quoting.openQuote || !next.nameStart) out.out(out.quote)
                    next
                }

        override fun print(current: CharClass, out: NamePrinter, next: CharClass, codePoint: CodePoint) {
            // Never prints anything
        }

    }

    private object NameSeparatorPrinter : NameCharPrinter {

        override fun append(current: CharClass, out: NamePrinter, next: CharClass) =
            if (next.separating) current // Skip subsequent separators
            else {
                // Separate next with space, unless it glues with previous one
                if (!out.lastNonSeparating.nameGluesWith(next)) out.out(SPACE)
                next
            }

        override fun print(current: CharClass, out: NamePrinter, next: CharClass, codePoint: CodePoint) {
            // Never prints anything
        }

    }

    private object NameBodyPrinter : NameCharPrinter {

        override fun append(current: CharClass, out: NamePrinter, next: CharClass) = next

        override fun print(current: CharClass, out: NamePrinter, next: CharClass, codePoint: CodePoint) {
            out.out(codePoint)
        }

    }

    private object NameEscapePrinter : NameCharPrinter {

        override fun append(current: CharClass, out: NamePrinter, next: CharClass) = next

        override fun print(current: CharClass, out: NamePrinter, next: CharClass, codePoint: CodePoint) {
            out.out(BACKSLASH)
            out.out(codePoint)
        }

    }

    private object NameCodePrinter : NameCharPrinter {

        override fun append(current: CharClass, out: NamePrinter, next: CharClass) = next

        override fun print(current: CharClass, out: NamePrinter, next: CharClass, codePoint: CodePoint) {
            out.out(BACKSLASH)
            Integer.toHexString(codePoint).forEach { out.out(it.toInt()) }
            out.out(BACKSLASH)
        }

    }

    companion object {

        @JvmStatic
        fun of(codePoint: CodePoint): CharClass = when {
            Character.isWhitespace(codePoint) -> SEPARATOR_CHAR_CLASS
            OPERATORS.get(codePoint) -> OPERATOR_CHAR_CLASS
            QUOTES.get(codePoint) -> QUOTE_CHAR_CLASS
            BUILTINS.get(codePoint) -> BUILTIN_CHAR_CLASS
            codePoint == BACKSLASH -> ESCAPE_CHAR_CLASS
            else -> when (Character.getType(codePoint)) {
                in Character.UPPERCASE_LETTER..Character.OTHER_LETTER -> LETTER_CHAR_CLASS
                in Character.DECIMAL_DIGIT_NUMBER..Character.LETTER_NUMBER -> NUMERIC_CHAR_CLASS
                in Character.MATH_SYMBOL..Character.OTHER_SYMBOL -> OPERATOR_CHAR_CLASS
                Character.DASH_PUNCTUATION.toInt(), Character.CONNECTOR_PUNCTUATION.toInt() -> CONNECTOR_CHAR_CLASS
                in Character.START_PUNCTUATION..Character.OTHER_PUNCTUATION -> PUNCTUATION_CHAR_CLASS
                in Character.SPACE_SEPARATOR..Character.PARAGRAPH_SEPARATOR -> SEPARATOR_CHAR_CLASS
                else -> OTHER_CHAR_CLASS
            }
        }

    }

}
